android/databases/src/main/java/com/coinbase/wallet/libraries/databases/interfaces/DatabaseDaoInterface.kt

```diff
@@ -0,0 +1,66 @@
+package com.coinbase.wallet.libraries.databases.interfaces
+
+import androidx.room.Delete
+import androidx.room.Insert
+import androidx.room.OnConflictStrategy
+import androidx.room.RawQuery
+import androidx.room.Update
+import androidx.sqlite.db.SupportSQLiteQuery
+import io.reactivex.Completable
+import io.reactivex.Single
+
+/**
+ * DAO interface used as a common DAO interface in Database object
+ */
+interface DatabaseDaoInterface<T> {
+    /**
+     * Adds new models to the database.
+     *
+     * @param models The models to add to the database.
+     *
+     * @return A Completable indicating whether the operation completed
+     */
+    @Insert(onConflict = OnConflictStrategy.ABORT)
+    fun add(model: List<T>): Completable
+
+    /**
+     * Adds new models or update if existing records are found.
+     *
+     * @param models The models to add to the database.
+     *
+     * @return A Completable indicating whether the operation completed
+     */
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    fun addOrUpdate(model: List<T>): Completable
+
+    /**
+     * Updates the objects in the datastore
+     *
+     * @param models The objects to update in the database
+     *
+     * @return A Completable indicating whether the operation completed
+     */
+    @Update
+    fun update(model: List<T>): Completable
+
+    /**
+     * Fetches objects from the data store using given raw SQL
+     *
+     * @param query SQL query used to fetch the data
+     * @param args Argument passed to fill placeholders in the query above
+     *
+     * @return A Single wrapping the items returned by the fetch.
+     */
+    @RawQuery
```

Comment:
We probably need to annotate this with @Transaction.  Otherwise, this could return corrupt date for large queries. https://developer.android.com/reference/android/arch/persistence/room/Transaction

Reply:
done